src/xb-arena.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2025 Owen Chiaventone <ochiaventone@gmail.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#pragma once
+
+/* Simple arena allocator used to construct nodes during parsing pass
+ * If the nodes and strings on them were individually heap allocated instead, we would churn the
+ * heap with a large number of small and medium sized allocations. This is bad for performance in
+ * a short lived program, but more importantly can trigger serious heap fragmentation if you parse
+ * a large XML in the middle of a long-lived program.
+ * This was implicated in https://gitlab.gnome.org/GNOME/gnome-software/-/issues/941
+ *
+ * We don't know how large the XML is when we start parsing, so the arena may have to grow.
+ * This arena allocator uses a chunking strategy - if we run out, allocate another large block. 
+ */
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+/* default to 1MB chunks for arena allocation */
+#define XB_ARENA_CHUNKSIZE 1<<20
+
+typedef struct {
+    uint32_t n_chunks;
+    uint32_t chunks_cap;
+    /* Array of pointers to chunks */
+    uint8_t** chunks;
+    /* Points to first free byte in current chunk */
+    uint8_t* tail;
+    uint8_t* current_chunk_end;
+} XbArena;
+
+XbArena xb_arena_new(void);
+
+// Allocate bytes from the arena
```

Comment:
Similarly these comments would all move to being gtk-doc comments immediately above the function implementations in the C file.

Grep the rest of the source tree for `/**` for examples.